relay/metadata_manager.go

```diff
@@ -0,0 +1,154 @@
+	// the fragment size used for uploading the encoded chunks
+	fragmentSizeBytes uint32
+}
+
+// metadataManager encapsulates logic for fetching metadata for blobs. Utilized by the relay Server.
+type metadataManager struct {
+	ctx    context.Context
+	logger logging.Logger
+
+	// metadataStore can be used to read blob metadata from dynamoDB.
+	metadataStore *blobstore.BlobMetadataStore
+
+	// metadataCache is an LRU cache of blob metadata. Blobs that do not belong to one of the relay shards
+	// assigned to this server will not be in the cache.
+	metadataCache cache.CachedAccessor[v2.BlobKey, blobMetadata]
+
+	// shardSet is the set of shards assigned to this relay. This relay will refuse to serve metadata for blobs
+	// that are not assigned to one of these shards.
+	shardSet map[v2.RelayKey]struct{}
+
+	// pool is a work pool for managing concurrent metadata requests. Used to limit the number of concurrent
+	// requests to the metadata store.
+	pool *errgroup.Group
+}
+
+// newMetadataManager creates a new metadataManager.
+func newMetadataManager(
+	ctx context.Context,
+	logger logging.Logger,
+	metadataStore *blobstore.BlobMetadataStore,
+	metadataCacheSize int,
+	workPoolSize int,
+	shards []v2.RelayKey) (*metadataManager, error) {
+
+	shardSet := make(map[v2.RelayKey]struct{}, len(shards))
+	for _, shard := range shards {
+		shardSet[shard] = struct{}{}
+	}
+
+	pool, _ := errgroup.WithContext(ctx)
+	pool.SetLimit(workPoolSize)
+
+	server := &metadataManager{
+		ctx:           ctx,
+		logger:        logger,
+		metadataStore: metadataStore,
+		shardSet:      shardSet,
+		pool:          pool,
+	}
+
+	metadataCache, err := cache.NewCachedAccessor[v2.BlobKey, blobMetadata](metadataCacheSize, server.fetchMetadata)
+	if err != nil {
+		return nil, fmt.Errorf("error creating metadata cache: %w", err)
+	}
+
+	server.metadataCache = metadataCache
+
+	return server, nil
+}
+
+// metadataMap is a map of blob keys to metadata.
+type metadataMap map[v2.BlobKey]*blobMetadata
+
+// GetMetadataForBlobs retrieves metadata about multiple blobs in parallel.
+func (m *metadataManager) GetMetadataForBlobs(keys []v2.BlobKey) (*metadataMap, error) {
+
+	mapLock := sync.Mutex{}
+	mMap := make(metadataMap)
+	hadError := atomic.Bool{}
+
+	wg := sync.WaitGroup{}
+	wg.Add(len(keys))
+
+	for _, key := range keys {
+		boundKey := key
+		m.pool.Go(func() error {
+			defer wg.Done()
+
+			metadata, err := m.metadataCache.Get(boundKey)
+			if err != nil {
+				// Intentionally log at debug level. External users can force this condition to trigger
+				// by requesting metadata for a blob that does not exist, and so it's important to avoid
+				// allowing hooligans to spam the logs in production environments.
+				m.logger.Debug("error retrieving metadata for blob %s: %v", boundKey.Hex(), err)
+				hadError.Store(true)
+				return nil
+			}
+			mapLock.Lock()
+			mMap[boundKey] = metadata
+			mapLock.Unlock()
+
+			return nil
+		})
+	}
+	wg.Wait()
+
+	if hadError.Load() {
```

Comment:
should it early terminate if any of the requests failed instead of waiting all requests to finish?